cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/ArtifactCache.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.tethering;
+
+import com.google.common.net.HttpHeaders;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.common.ArtifactNotFoundException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.internal.remote.RemoteAuthenticator;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.internal.app.worker.sidecar.AbstractArtifactLocalizer;
+import io.cdap.cdap.proto.id.ArtifactId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import javax.annotation.Nullable;
+
+/**
+ * ArtifactCache is responsible for fetching and caching artifacts from tethered peers. The HTTP
+ * endpoints are defined in {@link ArtifactCacheHttpHandlerInternal}.
+ *
+ * Artifacts will be cached using the following file structure:
+ * /DATA_DIRECTORY/peers/<peer-name>/artifacts/<namespace>/<artifact-name>/<artifact-version>/
+ * <last-modified-timestamp>.jar
+ *
+ */
+public class ArtifactCache extends AbstractArtifactLocalizer {
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactCache.class);
+
+  @Inject
+  public ArtifactCache(CConfiguration cConf)
```

Comment:
Remove `public` for injected constructor.

Reply:
fixed